modules/standard/Heap.chpl

```diff
@@ -0,0 +1,420 @@
+    }
+
+    /*
+      Returns `true` if this heap contains zero elements.
+
+      :return: `true` if this heap is empty.
+      :rtype: `bool`
+    */
+    proc isEmpty(): bool {
+      _enter();
+      var result = _data.isEmpty();
+      _leave();
+      return result;
+    }
+
+    /*
+      Return the top element in the heap.
+
+      :return: The top element in the heap
+      :rtype: `eltType`
+
+      .. note::
+        The default behaviour is to return the maximal element.
+        *Maximal* is defined by ``comparator``. If a ``reverseComparator`` is passed to ``init``,
+        the heap will return the minimal element.
+
+    */
+    proc top() {
+      if (isOwnedClass(eltType)) {
+        compilerError("top() method is not avaliable on a 'heap'",
+                      " with elements of an owned type, here: ",
+                      eltType: string);
+      }
+      _enter();
+      if (boundsChecking && isEmpty()) {
+        boundsCheckHalt("Called \"heap.top\" on an empty heap.");
+      }
+      var result = _data[0];
+      _leave();
+      return result;
+    }
+
+    /*
+      Wrapper of comparing elements
+    */
+    pragma "no doc"
+    proc _greater(x: eltType, y: eltType) {
+      return chpl_compare(x, y, comparator) > 0;
+    }
+
+    /*
+      helper procs to maintain the heap
+    */
+    pragma "no doc"
+    proc _heapify_up(in pos: int) {
+      while (pos) {
+        var parent = pos / 2;
+        if (_greater(_data[pos],_data[parent])) {
+          _data[parent] <=> _data[pos];
+          pos = parent;
+        }
+        else break;
+      }
+    }
+
+    pragma "no doc"
+    proc _heapify_down(in pos: int) {
+      while (pos < _data.size) {
+        // find the child node with greater value
+        var greaterChild = pos*2;
+        if (greaterChild >= _data.size) then break; // reach leaf node, break
+        if (greaterChild + 1 < _data.size) {
+          // if the right child node exists
+          if (_greater(_data[greaterChild+1],_data[greaterChild])) {
+            // if the right child is greater, then update the greaterChild
+            greaterChild += 1;
+          }
+        }
+        // if the greaterChild's value is greater than current node, then swap and continue
+        if (_greater(_data[greaterChild],_data[pos])) {
+          _data[greaterChild] <=> _data[pos];
+          pos = greaterChild;
+        }
+        else break;
+      }
+    }
+
+    pragma "no doc"
+    proc _push(in element: eltType)
+    lifetime this < element {
+      _data.append(element);
+      _heapify_up(_data.size-1);
+    }
+    /*
+      Push an element into the heap
+
+      :arg element: The element that will be pushed
+      :type element: `eltType`
+    */
+    proc push(in element: eltType)
+    lifetime this < element {
+      _enter();
+      _push(element);
+      _leave();
+    }
+
+    /*
+      Push elements of a list into the heap
+
+      :arg x: The list of which elements is to push.
```

Comment:
That sentence got a little muddled, as did the corresponding one in the push(array) function.

Reply:
What about `The list to push`?

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/da9b1697a92c96653feed9deedb5cdcac9f9a4c6 Not sure about this.

Reply:
I would say `The list of elements to push`.